src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,241 @@
+ */
+template <typename... DataTypes>
+class TableBuilder {
+ public:
+  // names is a boost::hana::tuple of strings defining column names, types is a list of equal length defining respective
+  // column types types may contain std::optional<?>, which will result in a nullable column, otherwise columns are not
+  // nullable
+  template <typename Names>
+  TableBuilder(size_t chunk_size, const boost::hana::tuple<DataTypes...>& types, const Names& names,
+               opossum::UseMvcc use_mvcc, size_t estimated_rows = 0)
+      : _use_mvcc(use_mvcc), _estimated_rows_per_chunk(std::min(estimated_rows, chunk_size)) {
+    BOOST_HANA_CONSTANT_ASSERT(boost::hana::size(names) == boost::hana::size(types));
+
+    // Iterate over the column types/names and create the columns.
+    auto column_definitions = TableColumnDefinitions{};
+    boost::hana::for_each(boost::hana::zip(names, types), [&](const auto& name_and_type) {
+      auto name = name_and_type[boost::hana::llong_c<0>];
+      auto type = name_and_type[boost::hana::llong_c<1>];
+
+      auto data_type = data_type_from_type<table_builder::get_value_type<decltype(type)>>();
+      auto is_nullable = table_builder::is_optional_v<decltype(type)>;
+
+      column_definitions.emplace_back(name, data_type, is_nullable);
+    });
+    _table = std::make_shared<Table>(column_definitions, TableType::Data, chunk_size, use_mvcc);
+
+    // Reserve some space in the vectors
+    boost::hana::for_each(_value_vectors, [&](auto& values) { values.reserve(_estimated_rows_per_chunk); });
+    boost::hana::for_each(_null_value_vectors, [&](auto& null_values) {
+      if constexpr (std::decay_t<decltype(null_values)>::has_value) {
+        null_values.value().reserve(_estimated_rows_per_chunk);
+      }
+    });
+  }
+
+  std::shared_ptr<Table> finish_table() {
+    if (_current_chunk_row_count() > 0) {
+      _emit_chunk();
+    }
+
+    return _table;
+  }
+
+  // Types == DataTypes, we deduce again to get universal references, so we can call with lvalues
+  template <typename... Types>
+  void append_row(Types&&... new_values) {
+    auto values_tuple = boost::hana::make_tuple(std::forward<decltype(new_values)>(new_values)...);
+    BOOST_HANA_CONSTANT_ASSERT(boost::hana::size(_value_vectors) == boost::hana::size(values_tuple));
+
+    // Create tuples ([&values0, &null_values0, value0], [&values1, &null_values1, value1], ...)
+    auto value_vectors_and_null_value_vectors_and_values = boost::hana::zip_with(
+        [](auto& values, auto& null_values, auto&& value) {
+          return boost::hana::make_tuple(std::reference_wrapper(values), std::reference_wrapper(null_values),
+                                         std::forward<decltype(value)>(value));
+        },
+        _value_vectors, _null_value_vectors, values_tuple);
+
+    // The reason why we can't do the for_each stuff in zip_with is the following:
```

Comment:
feel free to remove this comment, sprinkling hana docs about the libraries basic concepts into our code won't help much, imo

Reply:
There was a comment about that in the original table_builder and it reduces confusion about why we first zip_with and then for_each, so I think it should stay.